### ISSUE TITLE
how can I tell what database I?

### ISSUE THREAD

**Issue by gavinking** [quarkusio/quarkus#40922]
### Description

Quarkus reroutes all Hibernate's INFO-level logging to DEBUG level. This means that I get no confirmation at all about what database or Hibernate is connecting to without turning on `DEBUG` logging using `quarkus.log.category."org.hibernate".level=DEBUG`, which fills my log with hundreds of useless messages which I have to hunt through to find what I'm looking for.

This results in a surprisingly poor experience at development time. It's really hard to see, for example, wether setting:

```
quarkus.datasource.db-kind = postgresql
quarkus.datasource.username = hreact
quarkus.datasource.password = hreact
quarkus.datasource.jdbc.url = jdbc:postgresql://localhost:5432/hreact
```

actually had any effect at all, or whether I'm still just connecting to h2.

**Comment by gavinking** [comment]
> Quarkus reroutes all Hibernate's INFO-level logging to DEBUG level. 


Hahaha, no, sorry, my bad. What actually happened is that fairly recently @Sanne changed a very useful message which has always been logged at `INFO` level to `DEBUG` (in this commit https://github.com/hibernate/hibernate-orm/commit/af86f96f274608f912479c7d9f53c88d6a157649).

I have never really noticed the change, because in my projects I have a standard log4j.properties file which sets:

```
# Print the selected dialect information
logger.hibernate-dialect.name = org.hibernate.orm.dialect
logger.hibernate-dialect.level = debug
```

which apparently is there precisely to undo the effect of that change.

Now, usually when I'm developing Hibernate, I would not even care much about this particular log message because there are a couple of other messages coming from `DriverManagerConnectionProvider` which log the JDBC URL and JDBC driver class, so I'm always sure exactly what database I'm connecting to.

But Agroal is *much* less noisy, logging *nothing at all*. So I don't see:

- the JDBC URL,
- the dialect, nor
- the JDBC driver.

And I don't even know what database I'm connected to.

**Comment by DavideD** [comment]
Maybe it's not related to this issue, but Quarkus seems to only be able to select the correct dialect only for supported databases. It would be nice if quarkus would let us know when it's using the default dialect  version because the database is not supported,  and mention the property `quarkus.datasource.db-version` to override it.

**Comment by yrodiere** [comment]
You can certainly alter the extension to log a line looking something like:

```
Persistence unit '<default>' will connect to datasource 'something': jdbc:postgresql://localhost:5432/hreact. Assuming database version 14.0.
```

And append something like this if db-version is not set explicitly:

```
Persistence unit '<default>' will connect to datasource 'something': jdbc:postgresql://localhost:5432/hreact. Assuming database version 14.0. The database version defaults to the oldest supported value; if your database is more recent, set the build-time property `quarkus.datasource.db-version` to benefit from more features.
```

That being said, people in this project have strong opinions about logs on startup and have been actively hunting them down.
I personally don't mind.

**Comment by DavideD** [comment]
> It would be nice if quarkus would let us know when it's using the default dialect version because the database is not supported, and mention the property `quarkus.datasource.db-version` to override it.

I see now that Quarkus cannot select on its own the correct dialect + version.
I'm not sure why these properties aren't mandatory (at least for unsupported databases), but I guess adding a message specifying this on each start up could be redundant.

**Comment by yrodiere** [comment]
> I'm not sure why these properties aren't mandatory (at least for unsupported databases),

I'm pretty sure the dialect is mandatory for unsupported databases.

The version is always optional, and the reason is part upstream (ORM has default versions for all dialects anyway), part history (the concept of DB version only appeared in ORM 6) and part Quarkus "philosophy" ("thou shall not require any configuration for Quarkus to start").

To be clear, I _also_ think it only makes sense to force people to explicitly select a DB version. I agree with you. We mandate it for Elasticsearch in Hibernate Search, by the way.
In practice though:

1. The default behavior of using the oldest supported version kind of works, at least as long as you're not relying on new-ish features.
2. Requiring people to set the DB version explicitly in Quarkus would be a breaking change affecting _lots_ of applications.

... so I don't see it happening anytime soon.

Displaying the defaulted version on startup could be a nice first step; at least people would be able to notice there is something wrong in their config.

**Comment by DavideD** [comment]
Thanks for the clarification.

> Displaying the defaulted version on startup could be a nice first step; at least people would be able to notice there is something wrong in their config

Yes, I think this would be really helpful

**Comment by gavinking** [comment]
> You can certainly alter the extension to log a line looking something like:
> 
> ```
> Persistence unit '<default>' will connect to datasource 'something': jdbc:postgresql://localhost:5432/hreact. Assuming database version 14.0.
> ```
> 
> And append something like this if db-version is not set explicitly:
> 
> ```
> Persistence unit '<default>' will connect to datasource 'something': jdbc:postgresql://localhost:5432/hreact. Assuming database version 14.0. The database version defaults to the oldest supported value; if your database is more recent, set the build-time property `quarkus.datasource.db-version` to benefit from more features.
> ```

Yes this is all reasonable. I would also log the product name of the database FWIW.

**Comment by gavinking** [comment]
So what we still need to establish is, exactly who is responsible for logging what:

I would propose that we add `String getLoggableDescrption()` or something to `ConnectionProvider`, so that the pool can give us a description of what it's connected to.

But then:

- Does Hibernate itself log that? This is my preference, but I'm not sure how we control whether this message will be logged at build vs run time in Quarkus, and we would need to ensure that Quarkus is not hiding Hibernate's log messages.
- Or does the Quarkus Hibernate extension log it?
- Or both?

**Comment by yrodiere** [comment]
> Does Hibernate itself log that? This is my preference

Note this will prevent:

* Logging Quarkus-specific information (e.g. datasource name)
* Adapting the log to Quarkus-specific conditions ("using the default DB version") and logging Quarkus-specific hints (e.g. "your DB version isn't set, use `quarkus.something` to set it")

So even if Hibernate ORM does it, we might want to override what's done and do it again in Quarkus with more context.

> I'm not sure how we control whether this message will be logged at build vs run time in Quarkus

It basically depends where you do it in the build process. The later, the better as far as Quarkus is concerned. `JdbcEnvironmentInitiator` is *not* a good place as it happens before metadata creation (done at build time, or at least static init, in Quarkus); something that happens during `SessionFactory` creation would be a good place.

> and we would need to ensure that Quarkus is not hiding Hibernate's log messages.

Quarkus only hides very specific messages, so it's unlikely it will hide a new log.

**Comment by gavinking** [comment]
https://hibernate.atlassian.net/browse/HHH-18224

**Comment by gavinking** [comment]
@yrodiere so it sounds like "both" is viable...

**Comment by gavinking** [comment]
So in https://github.com/hibernate/hibernate-orm/pull/8525, @jrenaat did some work to improve the way Hibernate logs connection information. Prior to that I also cleaned up some stuff that was a bit more verbose than it needed to be,

But as of Quarkus 3.16.3, Quarkus is still suppressing the Hibernate logs, and I still don't see any information confirming that I'm connecting to the database I intend to connect to. (Nor is the info visible in the Dev UI.)

Look, I think the easiest solution here is for Quarkus to simply stop suppressing the Hibernate logs. That's a bad idea. Hibernate does not log a heap of irrelevant stuff at INFO level, it only logs things that I typically _do_ want to see.

**Comment by Sanne** [comment]
I agree that this old decision needs to be revisited.

For context, in early days, we decided to define simple to follow guidelines for all extensions : "thou shall not log anything at all".  Which was obviously  as idealistic as "you'll guess what is in user's mind and self-configure flawlessly".

In fairness, this was mostly to address other less-so polished libraries but needed Hibernate to set the example. I'd also argue that details are less crucial for many other libraries, but I might be biased :)

This really needs to evolve, though. I had the same bad experience multiple times.  cc/ @maxandersen

**Comment by gsmet** [comment]
> I'd also argue that details are less crucial for many other libraries, but I might be biased :)

That's exactly the problem as you are well aware :). Every library owner will think their information needs to be there and in the end you end up with a 200 lines startup - which is exactly what we collectively tried to avoid.
Why would Hibernate ORM logging be more important than the Kafka logging and the `<insert another component here>` logging?
We took this decision for good reasons at the beginning of the project. It's not an accident.

The other problem is that each component has its own sauce for logging at startup and you end up with an highly inconsistent startup - which is not ideal for a platform like Quarkus where we try to have a good developer experience and a consistent one (when we can).

We might want to rediscuss this but it's a lot more nuanced  than just enabling Hibernate ORM logging and something that needs to be discussed at the project level.

It could also be a global option that you enable if you feel like it. But even that doesn't solve the problems I mention above. DevEx and consistency are important for Quarkus.

Finally that's definitely something we should ask the users as they will be the most impacted.
If 90% of the users want the logging, then we know it makes sense to enable it by default, if 90% don't want it, we need to not enable it by default.

Anyway, as mentioned, it requires a discussion at the project level as it's a change in the policy we enforced from the start (and it was a conscious decision at that time).

**Comment by yrodiere** [comment]
@gavinking I'm not entirely sure the problem lies in the Quarkus policy only. I remember having to suppress the logs because they were simply incorrect in Quarkus, most likely logged twice (static init + runtime init) with the first one beeing too early, with no useful information. But then there were follow-ups to @jrenaat's PR that fixed a few things (no logging when booting offline, like in static init), so maybe that's no longer a problem.
Bottom line: we'd have to check what happens in Quarkus nowadays if we re-enable these logs. Including build-time logs in native build. Nothing blocking, of course, but something to keep in mind.

Regarding policy, I'd argue that Quarkus startup logs in dev mode are _already_ verbose when using dev services. So maybe that's something we could do in dev mode only, at least for a start -- since _that_ arguably would be consistent with the dev service logs. And I really mean arguably, because I fully expect arguments to ensue :)

I also agree a more global, consistent approach to logging connection info would be better -- e.g. a single block of logs with all services Quarkus will connect to, displayed on startup. But unless we find someone willing to actually do that (quite extensive) work -- on ORM, but also Elasticsearch, MongoDB, Kafka, REST clients, ... anything that connects to the outside, really -- and to deal with the security risks (hello, logging of JDBC URLs containing inline passwords...), I suspect requiring this from the start will just result in nothing being done at all. Maybe let's just agree on a format and SPI, and implement this gradually on extensions when maintainers have time?

That is, after someone posts their proposal to the quarkus-dev mailing list, since I think that's what Guillaume is after here :)

**Comment by Sanne** [comment]
> We might want to rediscuss this but it's a lot more nuanced than just enabling Hibernate ORM logging and something that needs to be discussed at the project level.

Right I totally agree - and sure Gavin and me are biased, well aware of that, which is way I said it :) And to bring it up at project level is why I pinged Max, someone needs to find some volunteers to explore it further.

It definitely is a nuanced issue and I understand why we did it this way in early days. And we need consistency and guidelines - perhaps and ARD - but we definitely need some way to summarize/highlight _selected_ details, especially when the lack of these makes for a bad user experience.

> I also agree a more global, consistent approach to logging connection info would be better -- e.g. a single block of logs with all services Quarkus will connect to, displayed on startup. 

+1
We need some form of SPI which Quarkus extensions can use to let users know of important information. It doesn't have to be a traditional log, it could be more structured and joyful (coloured?), perhaps have a nice interactive place on the Dev console as well.

**Comment by gavinking** [comment]
Hi, I reported this in May last year, and there has still been no action taken.

This behavior is absolutely hostile to anyone using Quarkus with Hibernate. As far as I can see, there's quite literally **no reasonable way** to get Quarkus to log information about the database it's connecting to, short of filling up the log with hundreds of useless `DEBUG`-level messages from Hibernate, in which case the log message I'm looking for gets lost anyway.

This is just Not Good Enough.

**Comment by gavinking** [comment]
See: https://hibernate.atlassian.net/browse/HHH-19611

And: https://github.com/hibernate/hibernate-orm/pull/10520

I just lost a whole day, that I'm never going to get back, futzing around demoting `DEBUG` level log messages to `TRACE` in Hibernate ORM, and I really have better things to do with my life.

To be clear: what Quarkus does here is completely obnoxious. There are standard way to suppress log messages, and they don't involve routing `INFO`-level messages to `DEBUG`. Hibernate never logs anything at `INFO` which isn't **extremely** useful to know, at least at development time.

Note: the argument that Hibernate should be on the same footing as every other library used by Quarkus is just silly. Most other libraries don't connect to databases, and the ones that do should be allowed to log stuff. I'm sorry, but there's an obvious material difference here between Hibernate ORM and, say, HIbernate Validator, or Jackson, or whatever.

**Comment by yrodiere** [comment]
> This is just Not Good Enough.

Guess what, this is What We Could Do. If it's not enough, tough luck. Other things have been worked on, when someone was working on the Quarkus Hibernate extensions, and this didn't didn't make it to the top of the list. 

Progress on this will need someone available to work on it. Maybe @lucamolteni would be?

I'd suggest two steps:

1. Just make the Hibernate ORM log work. From what I remember, it current doesn't in Quarkus, at least not in native image, maybe in other cases. Enable it only in dev mode if that's necessary to get the PR merged.
2. Later, champion the more complex idea of a global log for external services a Quarkus app connects to -- see https://github.com/quarkusio/quarkus/issues/40922#issuecomment-2490603451

**Comment by lucamolteni** [comment]
> Progress on this will need someone available to work on it. Maybe [@lucamolteni](https://github.com/lucamolteni) would be?

Yes I can start working on this in the following weeks

**Comment by jrenaat** [comment]
> Just make the Hibernate ORM log work.
IIRC I did some change in ORM a while back for quarkus' sake, but I forgot what that was related to ...

**Comment by beikov** [comment]
Can we close this now? We have logging in ORM 7.1 which is part of Quarkus already, so all good, right?

**Comment by yrodiere** [comment]
No, not all good AFAIK. Last time I checked the logging on Quarkus was very lacking, due to how Quarkus integrates with Hibernate ORM. Someone needs to try and report what's wrong / can be improved exactly.

Regarding the how... as I mentioned above:


> I'd suggest two steps:
> 
>     1. Just make the Hibernate ORM log work. From what I remember, it current doesn't in Quarkus, at least not in native image, maybe in other cases. Enable it only in dev mode if that's necessary to get the PR merged.
> 
>     2. Later, champion the more complex idea of a global log for external services a Quarkus app connects to -- see

**Comment by FroMage** [comment]
I can't find back where that discussion happened, but we did have a long discussion about what info should or not be logged by Quarkus at startup, and I definitely agree that we must log the databases we connect to at startup. Ideally one line per DB, with DB name, type, port, user, namespace, possibly driver and DB version.
Those are all useful information.

The discussion recounted similar info for all systems we connect to, and attempted to present them all in a coherent format.

This should also be visible in the DEV UI if not already.
